File: go/bundle.go
package spiffile

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"sync"
)

// ErrUnknownIdentity is wrapped by KeysFor when the claimed identity has no
// keys in the trust bundle.
var ErrUnknownIdentity = fmt.Errorf("identity has no keys in trust bundle")

// ParseBundle parses and validates a trust bundle document.
//
// A malformed entry must not take down verification for everyone else in
// the bundle: invalid entries are skipped and reported as warnings rather
// than failing the parse — same semantics as every other spiffile
// implementation. Top-level malformation (missing members, wrong version)
// still fails.
func ParseBundle(doc []byte) (Bundle, []string, error) {
	var raw struct {
		TrustDomain     *string                    `json:"trust_domain"`
		SpiffileVersion *int                       `json:"spiffile_version"`
		Identities      map[string]json.RawMessage `json:"identities"`
	}
	if err := json.Unmarshal(doc, &raw); err != nil {
		return Bundle{}, nil, fmt.Errorf("bundle is not valid JSON: %w", err)
	}
	if raw.TrustDomain == nil || raw.SpiffileVersion == nil || raw.Identities == nil {
		return Bundle{}, nil, fmt.Errorf("bundle is missing a required member (trust_domain, spiffile_version, identities)")
	}
	if *raw.SpiffileVersion != BundleVersion {
		return Bundle{}, nil, fmt.Errorf("unsupported spiffile_version: %d", *raw.SpiffileVersion)
	}

	bundle := Bundle{
		TrustDomain:     *raw.TrustDomain,
		SpiffileVersion: *raw.SpiffileVersion,
		Identities:      map[string]KeySet{},
	}
	var warnings []string
	for rawID, rawKeys := range raw.Identities {
		id, err := ParseSpiffeID(rawID)
		if err != nil {
			warnings = append(warnings, fmt.Sprintf("skipping invalid bundle entry %q: %v", rawID, err))
			continue
		}
		if id.TrustDomain != bundle.TrustDomain {
			warnings = append(warnings, fmt.Sprintf("skipping bundle entry %q: not in trust domain %q", rawID, bundle.TrustDomain))
			continue
		}
		var keySet KeySet
		if err := json.Unmarshal(rawKeys, &keySet); err != nil || len(keySet.Keys) == 0 {
			warnings = append(warnings, fmt.Sprintf("skipping bundle entry %q: no keys", rawID))
			continue
		}
		bundle.Identities[id.String()] = keySet
	}
	return bundle, warnings, nil
}

// KeysFor returns the JWKs bound to one identity.
func (b Bundle) KeysFor(spiffeID string) ([]json.RawMessage, error) {
	keySet, ok := b.Identities[spiffeID]
	if !ok {
		return nil, fmt.Errorf("%w: %q", ErrUnknownIdentity, spiffeID)
	}
	return keySet.Keys, nil
}

// FileBundleSource reads a bundle from a file, hot-reloading when the file
// changes (mtime+size), so secret-sync style updates are picked up without
// restarts. Parse warnings are logged via slog once per reload.
type FileBundleSource struct {
	path string

	mu     sync.Mutex
	mtime  int64
	size   int64
	bundle *Bundle
}

// NewFileBundleSource creates a source for the bundle at path.
func NewFileBundleSource(path string) *FileBundleSource {
	return &FileBundleSource{path: path}
}

// Get returns the current bundle, re-reading the file if it changed.
func (s *FileBundleSource) Get() (Bundle, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	info, err := os.Stat(s.path)
	if err != nil {
		return Bundle{}, fmt.Errorf("cannot stat bundle at %s: %w", s.path, err)
	}
	if s.bundle != nil && info.ModTime().UnixNano() == s.mtime && info.Size() == s.size {
		return *s.bundle, nil
	}

	doc, err := os.ReadFile(s.path)
	if err != nil {
		return Bundle{}, fmt.Errorf("cannot read bundle at %s: %w", s.path, err)
	}
	bundle, warnings, err := ParseBundle(doc)
	if err != nil {
		return Bundle{}, fmt.Errorf("cannot parse bundle at %s: %w", s.path, err)
	}
	for _, w := range warnings {
		slog.Warn("spiffile bundle", "warning", w, "path", s.path)
	}
	s.bundle = &bundle
	s.mtime = info.ModTime().UnixNano()
	s.size = info.Size()
	return bundle, nil
}

File: ts/src/bundle.ts
/**
 * Trust bundle: the document binding SPIFFE IDs to their public keys.
 * Per-identity binding is what makes the self-issued model safe — a key may
 * only validate tokens whose `sub` it is bound to. See PROFILE.md.
 */
import { readFileSync, statSync } from "node:fs"

import { InvalidBundleError, InvalidSpiffeIdError, UnknownIdentityError } from "./errors.js"
import { SpiffeId } from "./spiffeId.js"
import type { Jwk } from "./keys.js"

export const BUNDLE_VERSION = 1

export type BundleDocument = {
  trust_domain: string
  spiffile_version: number
  identities: Record<string, { keys: Jwk[] }>
}

const warn: (message: string) => void =
  typeof process !== "undefined" && process.emitWarning
    ? (message) => process.emitWarning(message, "SpiffileBundleWarning")
    : console.warn

export class Bundle {
  readonly trustDomain: string
  readonly identities: ReadonlyMap<string, Jwk[]>

  constructor(trustDomain: string, identities: Map<string, Jwk[]>) {
    this.trustDomain = trustDomain
    this.identities = identities
  }

  static fromDocument(doc: unknown): Bundle {
    if (typeof doc !== "object" || doc === null) {
      throw new InvalidBundleError("bundle is not an object")
    }
    const { trust_domain, spiffile_version, identities } = doc as Partial<BundleDocument>
    if (typeof trust_domain !== "string" || spiffile_version === undefined || typeof identities !== "object" || identities === null) {
      throw new InvalidBundleError("bundle is missing a required member (trust_domain, spiffile_version, identities)")
    }
    if (spiffile_version !== BUNDLE_VERSION) {
      throw new InvalidBundleError(`unsupported spiffile_version: ${spiffile_version}`)
    }

    // A malformed entry must not take down verification for everyone else
    // in the bundle: skip it (loudly) instead of failing the parse.
    const parsed = new Map<string, Jwk[]>()
    for (const [rawId, jwks] of Object.entries(identities)) {
      let spiffeId: SpiffeId
      try {
        spiffeId = SpiffeId.parse(rawId)
      } catch (error) {
        if (error instanceof InvalidSpiffeIdError) {
          warn(`skipping invalid bundle entry '${rawId}': ${error.message}`)
          continue
        }
        throw error
      }
      if (spiffeId.trustDomain !== trust_domain) {
        warn(`skipping bundle entry '${rawId}': not in trust domain '${trust_domain}'`)
        continue
      }
      const keys = jwks && typeof jwks === "object" ? jwks.keys : undefined
      if (!Array.isArray(keys) || keys.length === 0) {
        warn(`skipping bundle entry '${rawId}': no keys`)
        continue
      }
      parsed.set(spiffeId.toString(), keys)
    }
    return new Bundle(trust_domain, parsed)
  }

  toDocument(): BundleDocument {
    return {
      trust_domain: this.trustDomain,
      spiffile_version: BUNDLE_VERSION,
      identities: Object.fromEntries([...this.identities].map(([id, keys]) => [id, { keys }])),
    }
  }

  keysFor(spiffeId: string): Jwk[] {
    const keys = this.identities.get(spiffeId)
    if (!keys) {
      throw new UnknownIdentityError(`no keys in trust bundle for '${spiffeId}'`)
    }
    return keys
  }
}

/**
 * Reads a bundle from a file, hot-reloading when the file changes
 * (mtime+size), so secret-sync style updates are picked up without restarts.
 */
export class FileBundleSource {
  readonly path: string
  private stamp: string | null = null
  private bundle: Bundle | null = null

  constructor(path: string) {
    this.path = path
  }

  get(): Bundle {
    const stat = statSync(this.path)
    const stamp = `${stat.mtimeMs}:${stat.size}`
    if (this.bundle === null || stamp !== this.stamp) {
      let doc: unknown
      try {
        doc = JSON.parse(readFileSync(this.path, "utf8"))
      } catch (error) {
        throw new InvalidBundleError(`cannot read bundle at ${this.path}: ${error}`)
      }
      this.bundle = Bundle.fromDocument(doc)
      this.stamp = stamp
    }
    return this.bundle
  }
}
